File: sdk-typescript/src/builders.ts
/**
 * HandoffRail SDK — Fluent builder pattern for constructing packets and chains.
 *
 * Usage:
 * ```typescript
 * const packetData = new PacketBuilder()
 *   .to('billing-01', 'BillingBot', { framework: 'custom' })
 *   .from('sales-01', 'SalesBot', { framework: 'langchain' })
 *   .withSummary('Customer wants Business tier')
 *   .withPriority('high')
 *   .withTags(['upgrade', 'business'])
 *   .withConversation([{ role: 'user', content: 'I want to upgrade' }])
 *   .withDecision('Proceed', { rationale: 'Customer is eligible' })
 *   .withAction({ description: 'Process payment', assignee: 'billing-01', priority: 'high' })
 *   .withDependency({ id: 'stripe', type: 'api', description: 'Payment gateway' })
 *   .withHitl({ reason: 'High-value upgrade needs approval', question: 'Approve?' })
 *   .build();
 * ```
 *
 * @module
 */

import type {
  PacketCreate,
  ChainHandoffRequest,
  Metadata,
  PacketContext,
  AgentInfo,
  TargetAgentInfo,
  ContextEntry,
  Artifact,
  Decision,
  Actions,
  PendingAction,
  CompletedAction,
  FailedAction,
  Dependency,
  HitlCheckpoint,
  Priority,
  ConversationRole,
} from './models';

// ── Helper ────────────────────────────────────────────────────────────────

function asPriority(p: string | Priority): Priority {
  const valid: Priority[] = ['low', 'normal', 'high', 'critical'];
  if (valid.includes(p as Priority)) return p as Priority;
  return 'normal';
}

// ── PacketBuilder ─────────────────────────────────────────────────────────

/**
 * Fluent builder for constructing {@link PacketCreate} payloads.
 *
 * Each method returns `this` so calls can be chained. Call {@link build}
 * at the end to get a validated {@link PacketCreate} object.
 */
export class PacketBuilder {
  protected _sourceId?: string;
  protected _sourceName?: string;
  protected _sourceFramework?: string;
  protected _sourceVersion?: string;
  protected _targetId?: string;
  protected _targetName?: string;
  protected _targetFramework?: string;
  protected _summary?: string;
  protected _priority: Priority = 'normal';
  protected _tags: string[] = [];
  protected _conversationState: ContextEntry[] = [];
  protected _artifacts: Artifact[] = [];
  protected _custom: Record<string, unknown> = {};
  protected _decisions: Decision[] = [];
  protected _pendingActions: PendingAction[] = [];
  protected _completedActions: CompletedAction[] = [];
  protected _failedActions: FailedAction[] = [];
  protected _dependencies: Dependency[] = [];
  protected _hitl?: HitlCheckpoint;
  protected _parentPacketId?: string;

  // ── Agent setters ────────────────────────────────────────────────────

  /**
   * Set the source (sending) agent.
   */
  from(agentId: string, agentName: string, options?: { framework?: string; version?: string }): this {
    this._sourceId = agentId;
    this._sourceName = agentName;
    this._sourceFramework = options?.framework;
    this._sourceVersion = options?.version;
    return this;
  }

  /**
   * Set the target (receiving) agent.
   */
  to(agentId: string, agentName: string, options?: { framework?: string }): this {
    this._targetId = agentId;
    this._targetName = agentName;
    this._targetFramework = options?.framework;
    return this;
  }

  // ── Context setters ──────────────────────────────────────────────────

  /** Set the context summary. */
  withSummary(summary: string): this {
    this._summary = summary;
    return this;
  }

  /** Set the packet priority. */
  withPriority(priority: string | Priority): this {
    this._priority = asPriority(priority);
    return this;
  }

  /** Set tags for filtering. */
  withTags(tags: string[]): this {
    this._tags = [...tags];
    return this;
  }

  /** Set the conversation state from a list of message-like entries. */
  withConversation(messages: ContextEntry[]): this {
    this._conversationState = [...messages];
    return this;
  }

  /** Add a single conversation entry. */
  addConversationEntry(role: ConversationRole, content: string, metadata?: Record<string, unknown>): this {
    this._conversationState.push({ role, content, metadata });
    return this;
  }

  /** Set named artifacts produced during the session. */
  withArtifacts(artifacts: Artifact[]): this {
    this._artifacts = [...artifacts];
    return this;
  }

  /** Set framework-specific custom context fields. */
  withCustom(custom: Record<string, unknown>): this {
    this._custom = { ...custom };
    return this;
  }

  // ── Decision / Action / Dependency helpers ───────────────────────────

  /**
   * Add a decision to the packet. An auto-incrementing ID is assigned.
   */
  withDecision(
    decision: string,
    options?: {
      rationale?: string;
      alternatives?: string[];
      decided_by?: string;
    },
  ): this {
    const id = `d${this._decisions.length + 1}`;
    this._decisions.push({
      id,
      decision,
      rationale: options?.rationale ?? '',
      alternatives: options?.alternatives,
      decided_by: options?.decided_by,
    });
    return this;
  }

  /**
   * Add a pending action to the packet.
   */
  withAction(options: {
    description: string;
    assignee: string;
    priority?: string | Priority;
    depends_on?: string[];
    action_id?: string;
  }): this {
    const id = options.action_id ?? `a${this._pendingActions.length + 1}`;
    this._pendingActions.push({
      id,
      description: options.description,
      assignee: options.assignee,
      priority: options.priority ? asPriority(options.priority) : undefined,
      depends_on: options.depends_on,
    });
    return this;
  }

  /**
   * Add an external dependency.
   */
  withDependency(options: {
    id: string;
    type?: 'data' | 'api' | 'human_approval' | 'external_event' | 'resource';
    description?: string;
    status?: 'blocked' | 'available' | 'unknown';
    source?: string;
  }): this {
    this._dependencies.push({
      id: options.id,
      type: options.type ?? 'api',
      description: options.description ?? '',
      status: options.status ?? 'unknown',
      source: options.source,
    });
    return this;
  }

  // ── HITL ─────────────────────────────────────────────────────────────

  /**
   * Add a human-in-the-loop checkpoint.
   */
  withHitl(options: {
    reason: string;
    question?: string;
    options?: string[];
    timeout_seconds?: number;
    required?: boolean;
  }): this {
    this._hitl = {
      required: options.required ?? true,
      reason: options.reason,
      question: options.question,
      options: options.options,
      timeout_seconds: options.timeout_seconds,
    };
    return this;
  }

  // ── Parent ───────────────────────────────────────────────────────────

  /** Set the parent packet ID for chained handoffs. */
  withParent(parentPacketId: string): this {
    this._parentPacketId = parentPacketId;
    return this;
  }

  // ── Build ────────────────────────────────────────────────────────────

  /**
   * Validate accumulated data and return a {@link PacketCreate} object.
   *
   * @throws {Error} If required fields (source agent, target agent, summary) are missing.
   */
  build(): PacketCreate {
    if (!this._sourceId || !this._sourceName) {
      throw new Error('Source agent is required. Call .from() before .build().');
    }
    if (!this._targetId || !this._targetName) {
      throw new Error('Target agent is required. Call .to() before .build().');
    }
    if (!this._summary) {
      throw new Error('Summary is required. Call .withSummary() before .build().');
    }

    const sourceAgent: AgentInfo = {
      id: this._sourceId,
      name: this._sourceName,
      framework: this._sourceFramework,
      version: this._sourceVersion,
    };

    const targetAgent: TargetAgentInfo = {
      id: this._targetId,
      name: this._targetName,
      framework: this._targetFramework,
    };

    const metadata: Metadata = {
      source_agent: sourceAgent,
      target_agent: targetAgent,
      priority: this._priority,
      tags: this._tags,
    };

    const context: PacketContext = {
      summary: this._summary,
      conversation_state: this._conversationState.length > 0 ? this._conversationState : undefined,
      artifacts: this._artifacts.length > 0 ? this._artifacts : undefined,
      custom: Object.keys(this._custom).length > 0 ? this._custom : undefined,
    };

    const actions: Actions = {
      pending: this._pendingActions.length > 0 ? this._pendingActions : undefined,
      completed: this._completedActions.length > 0 ? this._completedActions : undefined,
      failed: this._failedActions.length > 0 ? this._failedActions : undefined,
    };

    // Only include actions if at least one list is non-empty
    const actionsMaybe = actions.pending || actions.completed || actions.failed ? actions : undefined;

    return {
      parent_packet_id: this._parentPacketId,
      metadata,
      context,
      decisions: this._decisions.length > 0 ? this._decisions : undefined,
      actions: actionsMaybe,
      dependencies: this._dependencies.length > 0 ? this._dependencies : undefined,
      hitl: this._hitl,
    };
  }
}

// ── ChainBuilder ──────────────────────────────────────────────────────────

/**
 * Fluent builder for constructing {@link ChainHandoffRequest} payloads.
 *
 * Similar to {@link PacketBuilder} but produces a {@link ChainHandoffRequest}
 * (no `parent_packet_id` — that is provided as a separate argument to
 * the client method).
 */
export class ChainBuilder {
  protected _sourceId?: string;
  protected _sourceName?: string;
  protected _sourceFramework?: string;
  protected _targetId?: string;
  protected _targetName?: string;
  protected _targetFramework?: string;
  protected _summary?: string;
  protected _priority: Priority = 'normal';
  protected _tags: string[] = [];
  protected _conversationState: ContextEntry[] = [];
  protected _artifacts: Artifact[] = [];
  protected _custom: Record<string, unknown> = {};
  protected _decisions: Decision[] = [];
  protected _pendingActions: PendingAction[] = [];
  protected _dependencies: Dependency[] = [];
  protected _hitl?: HitlCheckpoint;

  /** Set the source (sending) agent. */
  from(agentId: string, agentName: string, options?: { framework?: string }): this {
    this._sourceId = agentId;
    this._sourceName = agentName;
    this._sourceFramework = options?.framework;
    return this;
  }

  /** Set the target (receiving) agent. */
  to(agentId: string, agentName: string, options?: { framework?: string }): this {
    this._targetId = agentId;
    this._targetName = agentName;
    this._targetFramework = options?.framework;
    return this;
  }

  /** Set the context summary. */
  withSummary(summary: string): this {
    this._summary = summary;
    return this;
  }

  /** Set the packet priority. */
  withPriority(priority: string | Priority): this {
    this._priority = asPriority(priority);
    return this;
  }

  /** Set tags for filtering. */
  withTags(tags: string[]): this {
    this._tags = [...tags];
    return this;
  }

  /** Set the conversation state. */
  withConversation(messages: ContextEntry[]): this {
    this._conversationState = [...messages];
    return this;
  }

  /** Set artifacts. */
  withArtifacts(artifacts: Artifact[]): this {
    this._artifacts = [...artifacts];
    return this;
  }

  /** Set custom context. */
  withCustom(custom: Record<string, unknown>): this {
    this._custom = { ...custom };
    return this;
  }

  /** Add a decision. */
  withDecision(
    decision: string,
    options?: { rationale?: string; decided_by?: string },
  ): this {
    const id = `d${this._decisions.length + 1}`;
    this._decisions.push({
      id,
      decision,
      rationale: options?.rationale ?? '',
      decided_by: options?.decided_by,
    });
    return this;
  }

  /** Add a pending action. */
  withAction(options: { description: string; assignee: string; priority?: string | Priority }): this {
    const id = `a${this._pendingActions.length + 1}`;
    this._pendingActions.push({
      id,
      description: options.description,
      assignee: options.assignee,
      priority: options.priority ? asPriority(options.priority) : undefined,
    });
    return this;
  }

  /** Add an external dependency. */
  withDependency(options: { id: string; type?: 'data' | 'api' | 'human_approval' | 'external_event' | 'resource'; description?: string }): this {
    this._dependencies.push({
      id: options.id,
      type: options.type ?? 'api',
      description: options.description ?? '',
      status: 'unknown',
    });
    return this;
  }

  /** Add a HITL checkpoint. */
  withHitl(options: { reason: string; question?: string; options?: string[] }): this {
    this._hitl = {
      required: true,
      reason: options.reason,
      question: options.question,
      options: options.options,
    };
    return this;
  }

  /**
   * Validate and return a {@link ChainHandoffRequest}.
   *
   * @throws {Error} If required fields are missing.
   */
  build(): ChainHandoffRequest {
    if (!this._sourceId || !this._sourceName) {
      throw new Error('Source agent is required. Call .from() before .build().');
    }
    if (!this._targetId || !this._targetName) {
      throw new Error('Target agent is required. Call .to() before .build().');
    }
    if (!this._summary) {
      throw new Error('Summary is required. Call .withSummary() before .build().');
    }

    const metadata: Metadata = {
      source_agent: { id: this._sourceId, name: this._sourceName, framework: this._sourceFramework },
      target_agent: { id: this._targetId, name: this._targetName, framework: this._targetFramework },
      priority: this._priority,
      tags: this._tags,
    };

    const context: PacketContext = {
      summary: this._summary,
      conversation_state: this._conversationState.length > 0 ? this._conversationState : undefined,
      artifacts: this._artifacts.length > 0 ? this._artifacts : undefined,
      custom: Object.keys(this._custom).length > 0 ? this._custom : undefined,
    };

    const actions: Actions = {
      pending: this._pendingActions.length > 0 ? this._pendingActions : undefined,
    };

    return {
      metadata,
      context,
      decisions: this._decisions.length > 0 ? this._decisions : undefined,
      actions: actions.pending ? actions : undefined,
      dependencies: this._dependencies.length > 0 ? this._dependencies : undefined,
      hitl: this._hitl,
    };
  }
}

File: sdk/src/handoffrail/sdk/builders.py
"""HandoffRail SDK — Fluent builder pattern for constructing packets and chains.

Usage:
    packet_data = (
        PacketBuilder()
        .to("billing-01", "BillingBot", framework="crewai")
        .from_("sales-01", "SalesBot", framework="langchain")
        .with_summary("Customer wants Business tier")
        .with_priority("high")
        .with_tags(["upgrade", "business"])
        .with_conversation([{"role": "user", "content": "I want to upgrade"}])
        .with_decision("Proceed", rationale="Customer is eligible")
        .with_action("Process payment", assignee="billing-01", priority="high")
        .with_dependency("stripe", type="api", description="Payment gateway")
        .with_hitl(reason="High-value upgrade needs approval", question="Approve?")
        .build()
    )

    chain_data = (
        ChainBuilder(parent_packet_id="abc-123")
        .to("followup-01", "FollowUpBot")
        .from_("billing-01", "BillingBot")
        .with_summary("Follow up on upgrade")
        .build()
    )
"""

from __future__ import annotations

from typing import Any

from handoffrail.sdk.models import (
    Actions,
    AgentInfo,
    Artifact,
    ChainHandoffRequest,
    ContextEntry,
    Decision,
    Dependency,
    HitlCheckpoint,
    Metadata,
    PacketContext,
    PacketCreate,
    PendingAction,
    Priority,
    TargetAgentInfo,
)


class PacketBuilder:
    """Fluent builder for constructing PacketCreate payloads.

    Each method returns ``self`` so calls can be chained.  Call ``build()``
    at the end to get a validated :class:`PacketCreate` instance.
    """

    def __init__(self) -> None:
        self._source_id: str | None = None
        self._source_name: str | None = None
        self._source_framework: str | None = None
        self._source_version: str | None = None
        self._target_id: str | None = None
        self._target_name: str | None = None
        self._target_framework: str | None = None
        self._summary: str | None = None
        self._priority: Priority = Priority.normal
        self._tags: list[str] = []
        self._conversation_state: list[dict[str, Any]] = []
        self._artifacts: list[dict[str, Any]] = []
        self._custom: dict[str, Any] = {}
        self._decisions: list[dict[str, Any]] = []
        self._pending_actions: list[dict[str, Any]] = []
        self._completed_actions: list[dict[str, Any]] = []
        self._failed_actions: list[dict[str, Any]] = []
        self._dependencies: list[dict[str, Any]] = []
        self._hitl: dict[str, Any] | None = None
        self._parent_packet_id: str | None = None

    # ── Agent setters ─────────────────────────────────────────────────────

    def from_(
        self,
        agent_id: str,
        agent_name: str,
        *,
        framework: str | None = None,
        version: str | None = None,
    ) -> PacketBuilder:
        """Set the source (sending) agent."""
        self._source_id = agent_id
        self._source_name = agent_name
        self._source_framework = framework
        self._source_version = version
        return self

    def to(self, agent_id: str, agent_name: str, *, framework: str | None = None) -> PacketBuilder:
        """Set the target (receiving) agent."""
        self._target_id = agent_id
        self._target_name = agent_name
        self._target_framework = framework
        return self

    # ── Context setters ───────────────────────────────────────────────────

    def with_summary(self, summary: str) -> PacketBuilder:
        """Set the context summary."""
        self._summary = summary
        return self

    def with_priority(self, priority: str | Priority) -> PacketBuilder:
        """Set the packet priority ('low', 'normal', 'high', 'critical')."""
        if isinstance(priority, str):
            priority = Priority(priority)
        self._priority = priority
        return self

    def with_tags(self, tags: list[str]) -> PacketBuilder:
        """Set tags for filtering."""
        self._tags = list(tags)
        return self

    def with_conversation(self, messages: list[dict[str, Any]]) -> PacketBuilder:
        """Set the conversation state from a list of message dicts.

        Each dict should have at least ``role`` and ``content`` keys.
        """
        self._conversation_state = list(messages)
        return self

    def with_artifacts(self, artifacts: list[dict[str, Any]]) -> PacketBuilder:
        """Set named artifacts produced during the session."""
        self._artifacts = list(artifacts)
        return self

    def with_custom(self, custom: dict[str, Any]) -> PacketBuilder:
        """Set framework-specific custom context fields."""
        self._custom = dict(custom)
        return self

    # ── Decision / Action / Dependency helpers ────────────────────────────

    def with_decision(
        self,
        decision: str,
        *,
        rationale: str = "",
        alternatives: list[str] | None = None,
        decided_by: str | None = None,
    ) -> PacketBuilder:
        """Add a decision to the packet. An auto-incrementing ID is assigned."""
        decision_id = f"d{len(self._decisions) + 1}"
        entry: dict[str, Any] = {
            "id": decision_id,
            "decision": decision,
            "rationale": rationale,
        }
        if alternatives:
            entry["alternatives"] = alternatives
        if decided_by:
            entry["decided_by"] = decided_by
        self._decisions.append(entry)
        return self

    def with_action(
        self,
        description: str,
        *,
        assignee: str,
        priority: str | Priority | None = None,
        depends_on: list[str] | None = None,
        action_id: str | None = None,
    ) -> PacketBuilder:
        """Add a pending action to the packet."""
        aid = action_id or f"a{len(self._pending_actions) + 1}"
        entry: dict[str, Any] = {
            "id": aid,
            "description": description,
            "assignee": assignee,
        }
        if priority:
            entry["priority"] = priority.value if isinstance(priority, Priority) else priority
        if depends_on:
            entry["depends_on"] = depends_on
        self._pending_actions.append(entry)
        return self

    def with_dependency(
        self,
        dep_id: str,
        *,
        type: str = "api",
        description: str = "",
        status: str = "unknown",
        source: str | None = None,
    ) -> PacketBuilder:
        """Add an external dependency."""
        entry: dict[str, Any] = {
            "id": dep_id,
            "type": type,
            "description": description,
            "status": status,
        }
        if source:
            entry["source"] = source
        self._dependencies.append(entry)
        return self

    # ── HITL ──────────────────────────────────────────────────────────────

    def with_hitl(
        self,
        *,
        reason: str,
        question: str | None = None,
        options: list[str] | None = None,
        timeout_seconds: int | None = None,
    ) -> PacketBuilder:
        """Add a human-in-the-loop checkpoint."""
        self._hitl = {
            "required": True,
            "reason": reason,
        }
        if question:
            self._hitl["question"] = question
        if options:
            self._hitl["options"] = options
        if timeout_seconds is not None:
            self._hitl["timeout_seconds"] = timeout_seconds
        return self

    # ── Parent ─────────────────────────────────────────────────────────────

    def with_parent(self, parent_packet_id: str) -> PacketBuilder:
        """Set the parent packet ID for chained handoffs."""
        self._parent_packet_id = parent_packet_id
        return self

    # ── Build ──────────────────────────────────────────────────────────────

    def build(self) -> PacketCreate:
        """Validate accumulated data and return a PacketCreate instance.

        Raises:
            ValueError: If required fields (source agent, target agent, summary) are missing.
        """
        if self._source_id is None or self._source_name is None:
            msg = "Source agent is required. Call .from_() before .build()."
            raise ValueError(msg)
        if self._target_id is None or self._target_name is None:
            msg = "Target agent is required. Call .to() before .build()."
            raise ValueError(msg)
        if self._summary is None:
            msg = "Summary is required. Call .with_summary() before .build()."
            raise ValueError(msg)

        source_agent = AgentInfo(
            id=self._source_id,
            name=self._source_name,
            framework=self._source_framework,
            version=self._source_version,
        )
        target_agent = TargetAgentInfo(
            id=self._target_id,
            name=self._target_name,
            framework=self._target_framework,
        )
        metadata = Metadata(
            source_agent=source_agent,
            target_agent=target_agent,
            priority=self._priority,
            tags=self._tags,
        )

        conversation_state = [ContextEntry.from_dict(m) for m in self._conversation_state]
        artifacts = [Artifact.from_dict(a) for a in self._artifacts]

        context = PacketContext(
            summary=self._summary,
            conversation_state=conversation_state,
            artifacts=artifacts,
            custom=self._custom,
        )

        decisions = [Decision.from_dict(d) for d in self._decisions]
        actions = Actions(
            pending=[PendingAction.from_dict(a) for a in self._pending_actions],
            completed=[],
            failed=[],
        )
        dependencies = [Dependency.from_dict(d) for d in self._dependencies]
        hitl = HitlCheckpoint.from_dict(self._hitl) if self._hitl else None

        return PacketCreate(
            parent_packet_id=self._parent_packet_id,
            metadata=metadata,
            context=context,
            decisions=decisions,
            actions=actions,
            dependencies=dependencies,
            hitl=hitl,
        )


class ChainBuilder:
    """Fluent builder for constructing ChainHandoffRequest payloads.

    Similar to PacketBuilder but produces a ChainHandoffRequest (no
    ``parent_packet_id`` — that is provided as a separate argument to
    the client method).
    """

    def __init__(self) -> None:
        self._source_id: str | None = None
        self._source_name: str | None = None
        self._source_framework: str | None = None
        self._target_id: str | None = None
        self._target_name: str | None = None
        self._target_framework: str | None = None
        self._summary: str | None = None
        self._priority: Priority = Priority.normal
        self._tags: list[str] = []
        self._conversation_state: list[dict[str, Any]] = []
        self._artifacts: list[dict[str, Any]] = []
        self._custom: dict[str, Any] = {}
        self._decisions: list[dict[str, Any]] = []
        self._pending_actions: list[dict[str, Any]] = []
        self._dependencies: list[dict[str, Any]] = []
        self._hitl: dict[str, Any] | None = None

    def from_(self, agent_id: str, agent_name: str, *, framework: str | None = None) -> ChainBuilder:
        self._source_id = agent_id
        self._source_name = agent_name
        self._source_framework = framework
        return self

    def to(self, agent_id: str, agent_name: str, *, framework: str | None = None) -> ChainBuilder:
        self._target_id = agent_id
        self._target_name = agent_name
        self._target_framework = framework
        return self

    def with_summary(self, summary: str) -> ChainBuilder:
        self._summary = summary
        return self

    def with_priority(self, priority: str | Priority) -> ChainBuilder:
        if isinstance(priority, str):
            priority = Priority(priority)
        self._priority = priority
        return self

    def with_tags(self, tags: list[str]) -> ChainBuilder:
        self._tags = list(tags)
        return self

    def with_conversation(self, messages: list[dict[str, Any]]) -> ChainBuilder:
        self._conversation_state = list(messages)
        return self

    def with_decision(
        self,
        decision: str,
        *,
        rationale: str = "",
        decided_by: str | None = None,
    ) -> ChainBuilder:
        decision_id = f"d{len(self._decisions) + 1}"
        entry: dict[str, Any] = {
            "id": decision_id,
            "decision": decision,
            "rationale": rationale,
        }
        if decided_by:
            entry["decided_by"] = decided_by
        self._decisions.append(entry)
        return self

    def with_action(
        self,
        description: str,
        *,
        assignee: str,
        priority: str | Priority | None = None,
    ) -> ChainBuilder:
        aid = f"a{len(self._pending_actions) + 1}"
        entry: dict[str, Any] = {
            "id": aid,
            "description": description,
            "assignee": assignee,
        }
        if priority:
            entry["priority"] = priority.value if isinstance(priority, Priority) else priority
        self._pending_actions.append(entry)
        return self

    def with_dependency(
        self,
        dep_id: str,
        *,
        type: str = "api",
        description: str = "",
    ) -> ChainBuilder:
        entry = {"id": dep_id, "type": type, "description": description, "status": "unknown"}
        self._dependencies.append(entry)
        return self

    def with_hitl(
        self,
        *,
        reason: str,
        question: str | None = None,
        options: list[str] | None = None,
    ) -> ChainBuilder:
        self._hitl = {"required": True, "reason": reason}
        if question:
            self._hitl["question"] = question
        if options:
            self._hitl["options"] = options
        return self

    def build(self) -> ChainHandoffRequest:
        """Validate and return a ChainHandoffRequest."""
        if self._source_id is None or self._source_name is None:
            msg = "Source agent is required. Call .from_() before .build()."
            raise ValueError(msg)
        if self._target_id is None or self._target_name is None:
            msg = "Target agent is required. Call .to() before .build()."
            raise ValueError(msg)
        if self._summary is None:
            msg = "Summary is required. Call .with_summary() before .build()."
            raise ValueError(msg)

        source = AgentInfo(id=self._source_id, name=self._source_name, framework=self._source_framework)
        target = TargetAgentInfo(id=self._target_id, name=self._target_name, framework=self._target_framework)
        metadata = Metadata(source_agent=source, target_agent=target, priority=self._priority, tags=self._tags)

        context = PacketContext(
            summary=self._summary,
            conversation_state=[ContextEntry.from_dict(m) for m in self._conversation_state],
            artifacts=[Artifact.from_dict(a) for a in self._artifacts],
            custom=self._custom,
        )

        decisions = [Decision.from_dict(d) for d in self._decisions]
        actions = Actions(
            pending=[PendingAction.from_dict(a) for a in self._pending_actions],
            completed=[],
            failed=[],
        )
        dependencies = [Dependency.from_dict(d) for d in self._dependencies]
        hitl = HitlCheckpoint.from_dict(self._hitl) if self._hitl else None

        return ChainHandoffRequest(
            metadata=metadata,
            context=context,
            decisions=decisions,
            actions=actions,
            dependencies=dependencies,
            hitl=hitl,
        )
